bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/ReadAheadManager.java

```diff
@@ -0,0 +1,806 @@
+                conf.getLong(READ_AHEAD_TIMEOUT_MS, DEFAULT_READ_AHEAD_TIMEOUT_MS));
+    }
+
+    public ReadAheadManager(EntryLogger entryLogger, EntryLocationIndex entryLocationIndex,
+                            ReadCache cache, DbLedgerStorageStats dbLedgerStorageStats, boolean enableReadAheadAsync,
+                            boolean submitReadAheadTaskImmediately, int readAheadTaskPoolSize,
+                            int readAheadEntries, int readAheadBytes, double preTriggerReadAheadRatio,
+                            long readAheadTaskExpiredTimeMs, long readAheadTimeoutMs) {
+        // external assistant components assignment
+        this.entryLogger = entryLogger;
+        this.entryLocationIndex = entryLocationIndex;
+        this.cache = cache;
+
+        // metrics
+        this.dbLedgerStorageStats = dbLedgerStorageStats;
+
+        // mode
+        this.enableReadAheadAsync = enableReadAheadAsync;
+
+        // common parameters
+        this.readAheadEntries = readAheadEntries;
+        this.readAheadBytes = readAheadBytes;
+
+        if (enableReadAheadAsync) {
+
+            // configurable arguments
+            this.submitReadAheadTaskImmediately = submitReadAheadTaskImmediately;
+            this.preTriggerReadAheadRatio = preTriggerReadAheadRatio;
+
+            this.readAheadTaskExpiredTimeMs = readAheadTaskExpiredTimeMs;
+            this.readAheadTimeoutMs = readAheadTimeoutMs;
+
+            // core components initialization
+            readAheadExecutor = Executors.newFixedThreadPool(
+                    readAheadTaskPoolSize, new DefaultThreadFactory("read-ahead"));
+            cleanupExecutor = Executors.newSingleThreadScheduledExecutor(
+                    new DefaultThreadFactory("read-ahead-cleanup"));
+
+            pendingReadAheadPositions = new ConcurrentHashMap<>();
+            pendingDeletePositions = new ConcurrentLinkedQueue<>();
+
+            inProgressReadAheadTaskStatuses = new ConcurrentHashMap<>();
+            pendingDeleteReadAheadTaskStatuses = new ConcurrentLinkedQueue<>();
+
+            cleanupExecutor.scheduleAtFixedRate(
+                    SafeRunnable.safeRun(this::removeExpiredReadAheadTasks), 30, 30, TimeUnit.SECONDS);
+        }
+    }
+
+    /**
+     * Shutdown the thread pools used in async read-ahead mode.
+     */
+    public void shutdown() {
+        if (enableReadAheadAsync) {
+            this.readAheadExecutor.shutdown();
+            this.cleanupExecutor.shutdown();
+        }
+    }
+
+    /**
+     * Trigger read-ahead and return the corresponding entry.
+     *
+     * @param ledgerId
+     * @param entryId
+     * @return
+     * @throws IOException
+     */
+    public ByteBuf readEntry(long ledgerId, long entryId) throws IOException {
+        if (enableReadAheadAsync) {
+            return readEntryUnderAsyncReadAhead(ledgerId, entryId);
+        } else {
+            return readEntryUnderSyncReadAhead(ledgerId, entryId);
```

Comment:
@wuzhanpeng  could you divide this PR into several little PRs?, so that it's more easy to review.
1)first PR focusing on the reorganize of current sync mode read ahead : such as move `fillReadAheadCache` and related method to the `ReadAheadManager`
this PR doesn't introduce the new feature, just prepare for the new feature, and this PR should has no impact to the system;
2)following PR focusing on the functionality, 1 or more PR needed, for example core functionality, monitoring could be split into different PR.